### PR TITLE
Support setting the driver pod launching timeout.

### PR DESCRIPTION
**Pull request by lins05**
## What changes were proposed in this pull request?

Support setting the driver pod launching timeout through `"spark.kubernetes.driverLaunchTimeout"`, and increase the default value from 30s to 60s. The current value of 30s is kind of short for pulling the image from public docker registry plus the container/JVM start time.

## How was this patch tested?

Manual test.

### CODE REVIEW

**Comment by foxish** [comment]
I don't think `driverLaunchTimeout` adequately captures what the timeout really signifies. It is "local file upload timeout", and we should call it something like that IMO. WDYT @ash211 @mccheah

**Comment by mccheah** [comment]
Driver launch timeout seems fine, I wouldn't want to make the config point too verbose. If we could come up with a config name that's concise and that captures the fact that we're waiting to upload data and start running the application, that's fine, but if anything we come up with is too verbose then this will suffice.

**Comment by foxish** [comment]
I guess it isn't too bad a name. :) My concern was that except when uploading local jars, we should support a "fire and forget" type of launch mode, which would obviate the need for this timeout, and behave like the other schedulers in cluster mode. This is okay for now though.

**Comment by mccheah** [comment]
Actually launch timeout is the correct term since the future's completion is contingent on both the pod watch indicating that the pod is in "ready" status and that the application has been submitted.

**Comment by foxish** [comment]
Sounds good.
Maybe the future should be contingent on the pod entering the Pending/Ready/Running state in case we're not submitting anything from the local machine?

**Comment by mccheah** [comment]
Hm, how is this different from the current world? Do you mean having two futures and blocking on each in turn - block on the first future to get the pod into ready status and block on the second future for finishing the submission to the driver server?

**Comment by erikerlandson** [comment]
`fileStagingTimeout` ?

**Comment by lins05** [comment]
> Maybe the future should be contingent on the pod entering the Pending/Ready/Running state in case we're not submitting anything from the local machine?

For me the practical problem is it takes quite some time to pull the driver image from the public docker registry.  But it does sound like a good idea to output the detailed reason or phase of the pod when aborting the submit to k8s cluster, e.g. "the pod is running but the submit failed with error xyz", or "the pod has been pending for too long". Such details would help the user understand the reason of the failure and help locating/fixing the underlying problem.

What about merging this PR (maybe with another better name than `spark.kubernetes.driverLaunchTimeout`), and leave the more detailed error message described above for a future PR?

**Comment by foxish** [comment]
LGTM. I'm ok with merging this for now. @mccheah, do you concur?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 1]
```diff
@@ -424,7 +426,7 @@ private[spark] object Client extends Logging {
   private val DRIVER_LAUNCHER_CONTAINER_NAME = "spark-kubernetes-driver-launcher"
   private val SECURE_RANDOM = new SecureRandom()
   private val SPARK_SUBMISSION_SECRET_BASE_DIR = "/var/run/secrets/spark-submission"
-  private val LAUNCH_TIMEOUT_SECONDS = 30
+  private val LAUNCH_TIMEOUT_SECONDS = 60
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: we should rename this to `DEFAULT_LAUNCH_TIMEOUT_SECONDS` now that it's configurable

**Review by ash211 — CHANGES_REQUESTED** [review]

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala (hunk 1 above)
Comment (lins05) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: Make sense. I've updated it.

**Review by ash211 — APPROVED** [review]

**Comment by mccheah** [comment]
Looks good to me

**Pull request merged**